smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -0,0 +1,251 @@
+
+/**
+ * {@link ExtensionElement} that contains the UserTune. <br>
+ * Instance of UserTuneElement can be created using {@link Builder#build()}
+ * method.
+ */
+public final class UserTuneElement implements ExtensionElement {
+
+    public static final String NAMESPACE = "http://jabber.org/protocol/tune";
+    public static final String ELEMENT = "tune";
+
+    private final String artist;
+    private final UInt16 length;
+    private final Integer rating;
+    private final String source;
+    private final String title;
+    private final String track;
+    private final URI uri;
+
+    public static final UserTuneElement EMPTY_USER_TUNE = null;
+
+    private UserTuneElement(Builder builder) {
+        this.artist = builder.artist;
+        this.length = builder.length;
+        this.rating = builder.rating;
+        this.source = builder.source;
+        this.title = builder.title;
+        this.track = builder.track;
+        this.uri = builder.uri;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    @Override
+    public XmlStringBuilder toXML(XmlEnvironment xmlEnvironment) {
+        XmlStringBuilder xml = new XmlStringBuilder(this);
+        if (isEmptyUserTune()) {
+            return xml.closeEmptyElement();
+        }
+        xml.rightAngleBracket();
+        xml.optAppend(getUserTuneXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getUserTuneXML() {
+
+        XmlStringBuilder userTuneXML = new XmlStringBuilder();
+        userTuneXML.optElement("artist", artist);
+        userTuneXML.optElement("length", length);
+        userTuneXML.optElement("rating", rating);
+        userTuneXML.optElement("source", source);
+        userTuneXML.optElement("title", title);
+        userTuneXML.optElement("track", track);
+        userTuneXML.optElement("uri", uri);
+        return userTuneXML.toString();
+    }
+
+    private boolean isEmptyUserTune() {
+        return this.equals(EMPTY_USER_TUNE);
+    }
+
+    public static boolean hasUserTuneElement(Message message) {
+        return message.hasExtension(ELEMENT, NAMESPACE);
+    }
+
+    public static UserTuneElement from(Message message) {
+        return message.getExtension(ELEMENT, NAMESPACE);
+    }
+
+    @Override
+    public int hashCode() {
+        HashCode.Builder builder = HashCode.builder();
+        builder.append(this);
+        int result = builder.build();
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        EqualsUtil.EqualsComperator<UserTuneElement> comparator = new EqualsComperator<UserTuneElement>() {
+
+            @Override
+            public void compare(org.jivesoftware.smack.util.EqualsUtil.Builder equalsBuilder, UserTuneElement other) {
+                equalsBuilder.append(this, other);
```

Comment:
That will only compares `Object` instance of `this`. Which is not really what we want here, as the outcome would be the same as if the `equals()` method was not overridden.